Facet improvements

mistydemeo opened the pull request:
This branch contains several improvements to the Facet service and to code that uses it:
- The charts now add a facet if you click on a format in the graph; this isn't very useful now, but the eventual goal is to have that filter the transfer tree.
- The Facet service now generates and returns a unique ID for each passed facet; this makes it easier to unambiguously remove facets whose value isn't a string (for example, facets which use filter functions).
- The Facet service now provides a read-only array of all active facets, as a more convenient way to iterate over facets for display in templates.
- When adding a facet, it's possible to pass additional data to be included in the facet object in the facet array; this makes it easier to store additional data to be used in templates.
- The facet selector template and controller make use of these new features to display all current facets, allow individual facets to be removed, and allow more than one facet to be specified for a given field.

mistydemeo committed:
Visualizations: add facet when clicking format

mistydemeo committed:
Facet: disallow chaining

A future version of the Facet service will return unique IDs for each added facet in order to allow
it to be removed more easily in the future.

mistydemeo committed:
Facet: track a unique ID for each facet

This allows facets to be fetched or removed by ID
instead of value; this is useful when, for example,
removing facets whose values are functions.

mistydemeo committed:
Facet: track a flat list of every facet

This read-only list is useful for external consumers
who need a list of every facet.

mistydemeo committed:
Facet: allow attributes to be added to objects

mistydemeo committed:
Display selected facets

This is a first step to allowing multiple facets
for the same values.

mistydemeo committed:
Facet: allow multiple facets per field

mistydemeo commented on app/services/facet.service.js:
```diff
@@ -12,36 +14,73 @@
       }
-      this.facets[name].push(value);
-      return this; // return self so that calls can be chained
+      data.id = Math.random(); // lol
```
Probably want something more than a single `Math.random()`.

Hwesta commented:
:+1: 

Other option is having the caller provide the identifier for the facet (something semantically useful?) instead of the Facet service returning a synthetic one.  I did a WIP implementation at https://github.com/artefactual-labs/appraisal-tab/commit/d203a1e78d0902522aae178756f4a42c0fc1b6d9

mistydemeo committed:
Facet selectors: filter lists to only appropriate entries

mistydemeo committed:
Facet: allow ID to be specified up front

mistydemeo commented:
I split the difference - IDs are generated randomly if not specified, but can be passed as an optional argument.

I've also added a fix to the facet selection UI, which filters the facet selection dropdown boxes to only the values that make sense.

mistydemeo commented:
Updated with a more "proper" random ID generator.

Hwesta commented:
Looks reasonable! :+1:
